cache resource providers to reduce expensive api call

ms-henglu opened the pull request:
When running azurerm provider initialize, there're two `GET https://management.azure.com/subscriptions/{subscription_id}/providers?api-version=2016-02-01` calls, which are very time consuming, each can take about 7 seconds.

The two API calls are used to ensure resource providers are registered and enhance resource provider's name validation.

This PR cached the results of the first request, so even with enabling the above 2 features, there'll only be one API request.

After:
Run terraform plan command with service principal authentication
```
Executed in    9.19 secs 
```
There's only one `/providers` API call
<img width="802" alt="image" src="https://user-images.githubusercontent.com/79895375/212824663-930a32fb-824f-473a-95f0-3207f196d9d8.png">


Before:
Run terraform plan command with service principal authentication
```
Executed in    16.01 secs 
```
There're two `/providers` API call
<img width="802" alt="image" src="https://user-images.githubusercontent.com/79895375/212825173-9766da1d-52c4-40fb-95ab-1d1f9b8b9029.png">

tombuildsstuff commented on internal/clients/builder.go:
```diff
@@ -152,7 +152,10 @@ func Build(ctx context.Context, builder ClientBuilder) (*Client, error) {
-		resourceproviders.CacheSupportedProviders(ctx, client.Resource.ProvidersClient)
+		_, err := resourceproviders.CacheSupportedProviders(ctx, client.Resource.ProvidersClient)
+		if err != nil {
+			log.Printf("[DEBUG] error caching providers: %s", err)
```
`resource` providers

```suggestion
			log.Printf("[DEBUG] error caching resource providers: %+s", err)
```

tombuildsstuff commented on internal/provider/provider.go:
```diff
@@ -416,14 +416,13 @@ func providerConfigure(p *schema.Provider) schema.ConfigureContextFunc {
 			// List all the available providers and their registration state to avoid unnecessary
 			// requests. This also lets us check if the provider credentials are correct.
-			providerList, err := client.Resource.ProvidersClient.List(ctx, nil, "")
+			availableResourceProviders, err := resourceproviders.CacheSupportedProviders(ctx, client.Resource.ProvidersClient)
```
we should use the same approach we do for Cached Locations here?

ms-henglu replied:
Got it, but the location cache is part of the `go-azure-helper`, I can make another PR for it, but I think it will be a breaking change of `go-azure-helper`.

ms-henglu replied:
Hi @katbyte , not sure I followed, do you mean that we should move the resource provider cache logic to `go-azure-helper`?

katbyte replied:
yes? as tom suggested we should follow the same approach we used for location

mybayern1974 replied:
The move-cache-to-go-azure-helper has been made here: https://github.com/hashicorp/go-azure-helpers/pull/158. Please help review that and after that appreciate ms-henglu could continue here. @tombuildsstuff @katbyte

tombuildsstuff reviewed (changes requested):
hi @ms-henglu

Thanks for this PR - I've taken a look through and left some comments inline to make this consistent with how caching for the Locations are done, if we can make those changes so this becomes consistent then we should be able to get this merged.

Thanks!

ms-henglu commented:
Hi @tombuildsstuff ,

Thanks for the code review!

I changed the `CacheSupportedProviders` to `CachedSupportedProviders`, so it could not only be caching the resource providers, but returning them if they're already cached and caching them if not cached yet. The cached resource providers only contain the name and registration state, I've removed the raw http response.

Would you please take another look? thanks!

tombuildsstuff commented:
hey @ms-henglu 

Apologies for the delayed reply to this - it's taken a little time to review the PR's on `hashicorp/go-azure-helpers` (and so to get to this PR).

I spent some time digging into this one last week, and whilst this PR looks good - in retrospect it appears that we're actually going to need to move this logic from `hashicorp/go-azure-helpers` fully into the provider.

Whilst having this located within `hashicorp/go-azure-helpers` makes sense at the moment (since this is shared between AzureRM and Azure Stack) - leaving this within that repository means that we'll be unable to migrate this across to using `hashicorp/go-azure-sdk` (without a circular reference), so unfortunately this is actually going to need to live elsewhere.

Because this logic needs to be migrated across to using `hashicorp/go-azure-sdk` too, I hope you don't mind but I've opened #21695 which both moves Resource Provider Registration into this repository, adds this cache, migrates this across to using `hashicorp/go-azure-sdk` at API Version `2022-09-01` (from `2016-02-01`) and fixes a bug spotted during provider registration. There's an open question as to whether that logic belongs in this Provider or in `hashicorp/go-azure-sdk` (so that the Azure Stack provider can reuse this logic), but for now moving this into the Provider seems reasonable.

As such whilst I'd like to thank you for this contribution, I'm going to close this PR in favour of #21695 which implements this functionality using `hashicorp/go-azure-sdk`.

Thanks!
